cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.program;
+
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.ProgramRunStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Heartbeat thread to publish heart beat messages with notification type heart_beat
+ */
+public class ProgramRunHeartbeat implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunHeartbeat.class);
+
+  private final ProgramStatePublisher programStatePublisher;
+  private final Map<String, String> properties;
+  private final String programRunId;
+
+  ProgramRunHeartbeat(ProgramStatePublisher messagingProgramStatePublisher, Map<String, String> properties) {
+    this.programStatePublisher = messagingProgramStatePublisher;
+    this.properties = new HashMap<>(properties);
+    // could be called from RESUMING state, so update status to running for heartbeat
+    this.properties.put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.RUNNING.name());
```

Comment:
This won't be an issue if you inline the class, but you don't want to make assumptions like this when creating a separate class, as anybody can create this class. At a quick glance, it looks like it just publishes a notification with the given publisher and given properties. The fact that it modifies the properties is hidden which can easily lead to misuse